### DISCUSSION TITLE
Integration tests with tiny_http

### DISCUSSION THREAD

**Discussion by wpd** [tiny-http/tiny-http#216] [Q&A]
Are there any examples of performing integration tests with tiny_http?  I have the start of a project basically running... it serves 3 or 4 pages the way I want.  But now I want to refactor my code, add new pages, etc.. .and I would like to ensure that nothing that is working now changes (from the outside looking in).  Do you have any suggestions as to how I might do this?

--wpd

**Comment by wpd** [comment]
For what it's worth, I am using something like the following at the start of each of my integration tests.
```
if let Ok(server) = Server::new(ServerConfig {
            addr: (Ipv4Addr::UNSPECIFIED, port),
            ssl: None,
        }) {
    thread::spawn(move || { run_server(server); });
}
```
The intent (keeping in mind that I am still learning Rust) is that the first test will spin up the new server (successfully) on the specified `port` and spawn a thread to handle requests for that server.  Subsequent tests in the same process will fail to create the new server, so I don't spin up new threads, but they can continue to use the existing server.

I use `minreq` to issue requests to the server in my integration tests and parse the output to see that it's what I expect.

I think it is way cool that, when the tests exit, and the spawned thread is terminated, the server shuts down gracefully and I don't have to wait for the socket to timeout before I can use it again.  Nice job tiny-http author(s) and/or Rust!

**Comment by kolbma** [comment]
I use some `before()` and `after()` with `cargo r` process and ureq as HTTP client:

https://gist.github.com/kolbma/b8bd787568e543ccaf7abe7b5ba1b8ee
